File: app/src/main/java/com/tsaysoft/nfpachemidv3/ChemSpecial.java
package com.tsaysoft.nfpachemidv3;

/**
 * <h1>ChemSpecial</h1>
 * <h2>(Chemical Special Warning Category, NFPA 704)</h2>
 * Specifies the special properties identified by the NFPA 704 fire diamond.
 * <p>
 *     The NFPA 704-standardized properties are as follows:
 *     <ul>
 *         <li><code>OXIDIZER</code>: the compound is an oxidiser (OX)</li>
 *         <li><code>SIMPLE_ASPHYXIANT</code>: the compound is a simple asphyxiant gas (SA)</li>
 *         <li><code>WATER_REACT</code>: the compound reacts violently with water (<s>W</s>)</li>
 *     </ul>
 *     There are other symbols that are not NFPA 704-standardized. They are currently unsupported.
 * </p>
 *
 * @author Clay Tsay
 * @version 00.01.00
 *
 * @see <a href="https://en.wikipedia.org/wiki/NFPA_704#Codes">NFPA 704 - Wikipedia</a>
 */
public enum ChemSpecial {
    /**
     * Denotes that the chemical is an oxidizer.
     * <p>
     *     Examples:
     *     <ul>
     *         <li>KNO<sub>3</sub></li>
     *         <li>KClO<sub>3</sub></li>
     *     </ul>
     * </p>
     *
     * @since 00.01.00
     */
    OXIDIZER,

    /**
     * Denotes that the chemical is a simple asphyxiant gas.
     * <p>
     *     Examples:
     *     <ul>
     *         <li>N<sub>2</sub></li>
     *         <li>Ar</li>
     *     </ul>
     * </p>
     *
     * @since 00.01.00
     */
    SIMPLE_ASPHYXIANT,

    /**
     * Denotes that the chemical reacts with water.
     * <p>
     *     Examples:
     *     <ul>
     *         <li>CH<sub>3</sub>NCO (a.k.a. MIC)</li>
     *         <li>CaC<sub>2</sub></li>
     *     </ul>
     * </p>
     *
     * @since 00.01.00
     */
    WATER_REACT;

    /**
     * Converts a <code>ChemSpecial</code> to a <code>String</code>.
     * <p>
     *     e.g. <code>SIMPLE_ASPHYXIANT</code> becomes "<code>SIMPLE_ASPHYXIANT</code>".
     * </p>
     * @return the <code>ChemSpecial</code> converted into a <code>String</code>
     *
     * @since 00.02.00
     */
    @Override
    public String toString() {
        switch(this) {
            case OXIDIZER:
                return "OXIDIZER";
            case SIMPLE_ASPHYXIANT:
                return "SIMPLE_ASPHYXIANT";
            case WATER_REACT:
                return "WATER_REACT";
        }

        throw new RuntimeException("unable to convert ChemSpecial to String");
    }
}

File: app/src/main/java/com/tsaysoft/nfpachemidv3/ChemSpecial.kt
package com.tsaysoft.nfpachemidv3

/**
 * <h1>ChemSpecial</h1>
 * <h2>(Chemical Special Warning Category, NFPA 704)</h2>
 * Specifies the special properties identified by the NFPA 704 fire diamond.
 *
 * The NFPA 704-standardized properties are as follows:
 * - `OXIDIZER`: the compound is an oxidiser (OX)
 * - `SIMPLE_ASPHYXIANT`: the compound is a simple asphyxiant gas (SA)
 * - `WATER_REACT`: the compound reacts violently with water (~~W~~)
 *
 * There are other symbols that are not NFPA 704-standardized. They are currently
 * unsupported. See [source]()
 *
 * @author Clay Tsay
 * @version 00.01.00
 * @see [NFPA 704 - Wikipedia](https://en.wikipedia.org/wiki/NFPA_704.Codes)
 */
enum class ChemSpecial {
    /**
     * Denotes that the chemical is an oxidizer.
     *
     *
     * Examples:
     *
     *  * KNO<sub>3</sub>
     *  * KClO<sub>3</sub>
     *
     *
     *
     * @since 00.01.00
     */
    OXIDIZER,

    /**
     * Denotes that the chemical is a simple asphyxiant gas.
     *
     *
     * Examples:
     *
     *  * N<sub>2</sub>
     *  * Ar
     *
     *
     *
     * @since 00.01.00
     */
    SIMPLE_ASPHYXIANT,

    /**
     * Denotes that the chemical reacts with water.
     *
     *
     * Examples:
     *
     *  * CH<sub>3</sub>NCO (a.k.a. MIC)
     *  * CaC<sub>2</sub>
     *
     *
     *
     * @since 00.01.00
     */
    WATER_REACT;

    /**
     * Converts a `ChemSpecial` to a `String`.
     *
     *
     * e.g. `SIMPLE_ASPHYXIANT` becomes "`SIMPLE_ASPHYXIANT`".
     *
     * @return the `ChemSpecial` converted into a `String`
     *
     * @since 00.02.00
     */
    override fun toString(): String {
        when (this) {
            OXIDIZER -> return "OXIDIZER"
            SIMPLE_ASPHYXIANT -> return "SIMPLE_ASPHYXIANT"
            WATER_REACT -> return "WATER_REACT"
        }

        throw RuntimeException("unable to convert ChemSpecial to String")
    }

    // TODO: Determine if this is necessary, or if there is a better solution
    fun fromString(string: String): ChemSpecial {
        when (string) {
            "OXIDIZER" -> return OXIDIZER
            "SIMPLE_ASPHYXIANT" -> return SIMPLE_ASPHYXIANT
            "WATER_REACT" -> return WATER_REACT
        }

        throw RuntimeException("unable to convert ChemSpecial to String")
    }
}
